src/main/java/org/opensearch/knn/index/codec/nativeindex/NativeIndexWriter.java

```diff
@@ -0,0 +1,287 @@
+import static org.opensearch.knn.common.KNNConstants.MODEL_ID;
+import static org.opensearch.knn.common.KNNConstants.PARAMETERS;
+import static org.opensearch.knn.index.codec.util.KNNCodecUtil.buildEngineFileName;
+import static org.opensearch.knn.index.engine.faiss.Faiss.FAISS_BINARY_INDEX_DESCRIPTION_PREFIX;
+
+/**
+ * Writes KNN Index for a field in a segment. This is intended to be used for native engines
+ */
+@AllArgsConstructor
+@Log4j2
+public class NativeIndexWriter {
+    private static final Long CRC32_CHECKSUM_SANITY = 0xFFFFFFFF00000000L;
+
+    private final SegmentWriteState state;
+    private final FieldInfo fieldInfo;
+    private final NativeIndexBuildStrategy indexBuilder;
+    // TODO: Add quantization state as a member variable
+
+    /**
+     * Gets the correct writer type from fieldInfo
+     *
+     * @param fieldInfo
+     * @return correct NativeIndexWriter to make index specified in fieldInfo
+     */
+    public static NativeIndexWriter getWriter(final FieldInfo fieldInfo, SegmentWriteState state) {
+        // TODO: Fetch the quantization state here and pass it to NativeIndexWriter
+
+        final KNNEngine knnEngine = extractKNNEngine(fieldInfo);
+        boolean isTemplate = fieldInfo.attributes().containsKey(MODEL_ID);
+        boolean iterative = !isTemplate && KNNEngine.FAISS == knnEngine;
+        if (iterative) {
+            return new NativeIndexWriter(state, fieldInfo, MemOptimizedNativeIndexBuildStrategy.getInstance());
+        }
+        return new NativeIndexWriter(state, fieldInfo, BulkVectorTransferIndexBuildStrategy.getInstance());
+    }
+
+    private boolean buildAndWriteIndex(final KNNVectorValues<?> knnVectorValues) throws IOException {
+        if (knnVectorValues.totalLiveDocs() == 0) {
+            log.debug("No live docs for field " + fieldInfo.name);
+            return false;
+        }
+
+        final KNNEngine knnEngine = extractKNNEngine(fieldInfo);
+        final String engineFileName = buildEngineFileName(
+            state.segmentInfo.name,
+            knnEngine.getVersion(),
+            fieldInfo.name,
+            knnEngine.getExtension()
+        );
+        final String indexPath = Paths.get(
+            ((FSDirectory) (FilterDirectory.unwrap(state.directory))).getDirectory().toString(),
+            engineFileName
+        ).toString();
+        state.directory.createOutput(engineFileName, state.context).close();
+
+        final BuildIndexParams nativeIndexParams = indexParams(fieldInfo, indexPath, knnEngine);
+        boolean indexBuilt = indexBuilder.buildAndWriteIndex(nativeIndexParams, knnVectorValues);
+        writeFooter(indexPath, engineFileName, state);
+
+        return indexBuilt;
+    }
+
+    /**
+     * flushes the index
+     * @param knnVectorValues
+     * @return boolean to indicate if the index was written
+     * @throws IOException
+     */
+    public boolean flushIndex(final KNNVectorValues<?> knnVectorValues) throws IOException {
```

Comment:
FlushIndex and MergeIndex is doing most of the things same except the metrics emission , Can we Create common function like ProcessIndex and call this ProcessIndex from both function with extra parameter with functional Interface that takes which graph emission method to run